[Refactor] Update damage and heal functions in the `Pokemon` class

DayKev opened the pull request:
## What are the changes the user will see?
The damage number pop-up when healing is converted to damage (ie: due to Liquid Ooze) will no longer be green (which normally indicates healing).

## Why am I making these changes?
`Pokemon.damage` and `Pokemon.damageAndUpdate` had many optional parameters, and when changing them to use destructuring I noticed a bunch of other jank.

## What are the changes from a developer perspective?
- `Pokemon.damage` and `Pokemon.damageAndUpdate` now use destructuring for optional parameters.
- `Pokemon.damage` is now `protected`.
- `EnemyPokemon.damage` also updated to match the changes.
- `Pokemon.heal` now handles the heal number popup.
- `HpSplitAttr` simplified.
- Tsdocs were updated.
- Nesting reduced in `PostTurnStatusEffectPhase` and `WeatherEffectPhase` by using early returns.
- Bug in Present test fixed.
- All instances of `pokemon.turnData.damageTaken += ...` were consolidated into `Pokemon.damage()`.
- Added `ignoreDynamaxReduction` param to `Pokemon.damageAndUpdate` which causes the damage not to be reduced by the custom dynamax damage reduction factor if `true`.

## Checklist
- [x] Otherwise: **I'm using `beta` as my base branch**
- [x] There is no overlap with another PR?
- [x] The PR is self-contained and cannot be split into smaller PRs?
- [x] Have I provided a clear explanation of the changes?
- [x] Have I tested the changes manually?
- [x] Are all unit tests still passing? (`npm run test:silent`)
  - [x] Have I created new automated tests (`npm run test:create`) or updated existing tests related to the PR's changes?

DayKev committed:
[Refactor] Update damage and heal functions in the `Pokemon` class

`.damage` and `.damageAndUpdate` now use
destructuring for optional parameters.

`EnemyPokemon.damage` also updated to match.

`Pokemon.damage` is now `protected`.

`.heal` now handles the heal number popup.

Tsdocs were updated.

DayKev committed:
Simplify `HpSplitAttr`

DayKev committed:
Simplify `PokemonHealPhase`

DayKev committed:
Simplify `PostTurnStatusEffectPhase`

Use early returns to reduce nesting

Replace bitshifting with standard division

DayKev committed:
Organize imports in `pokemon-heal-phase.ts`

DayKev committed:
Organize imports in `post-turn-status-effect-phase.ts`

DayKev committed:
Update `WeatherEffectPhase`, reduce nesting with early returns

DayKev committed:
Update Present test and `test/testUtils/gameWrapper.ts`

DayKev committed:
Update `.damageAndUpdate` calls

DayKev committed:
Fix damage number display when healing is blocked

DayKev committed:
Rename param `critical` to `isCritical` in `damageAndUpdate`

DayKev committed:
Fix missing `end()` in `WeatherEffectPhase`

Tempo-anon commented on src/field/pokemon.ts:
```diff
@@ -3045,91 +3046,129 @@ export abstract class Pokemon extends Phaser.GameObjects.Container {
-   * @returns integer representing damage
+   * Deals damage to the pokemon. Called by {@linkcode damageAndUpdate}
+   * @param amount - Amount of damage that should be dealt
+   * @param ignoreSegments - Only used by {@linkcode EnemyPokemon.damage}
+   * @param preventEndure - If `true`, bypasses the effects of Endure, Sturdy, etc. Default `false`
+   * @param ignoreFaintPhase - If `true`, won't push a {@linkcode FaintPhase}. Default `false`
+   * @param ignoreDynamaxReduction - If `true`, dynamax damage reduction will be ignored. Default `false`
```
nit: the order of these should match the order of the params

DayKev replied:
They do.

Tempo-anon commented on src/field/pokemon.ts:
```diff
@@ -3045,91 +3046,129 @@ export abstract class Pokemon extends Phaser.GameObjects.Container {
-    ignoreFaintPhase: boolean = false,
+  protected damage(
+    amount: number,
+    {
+      preventEndure = false,
+      ignoreFaintPhase = false,
+      ignoreDynamaxReduction = false,
```
Is there a reason these 3 are now bundled together?

DayKev replied:
In hindsight it wasn't necessary to do this for `Pokemon.damage` since it's only supposed to be used by `Pokemon.damageAndUpdate`, but I've already done it so... 🤷

Tempo-anon commented on src/field/pokemon.ts:
```diff
@@ -3045,91 +3046,129 @@ export abstract class Pokemon extends Phaser.GameObjects.Container {
-    if (!preventEndure && this.hp - damage <= 0) {
+    const surviveDamage = new BooleanHolder(false);
+
+    if (!preventEndure && this.hp - amount <= 0) {
```
nit: Change `this.hp - amount <= 0` to `amount >= this.hp` or `this.hp <= amount` for simplicity

Tempo-anon commented on src/field/pokemon.ts:
```diff
@@ -3045,91 +3046,129 @@ export abstract class Pokemon extends Phaser.GameObjects.Container {
-   * @returns integer representing damage
+   * Deals damage to the pokemon. Called by {@linkcode damageAndUpdate}
+   * @param amount - Amount of damage that should be dealt
+   * @param ignoreSegments - Only used by {@linkcode EnemyPokemon.damage}
+   * @param preventEndure - If `true`, bypasses the effects of Endure, Sturdy, etc. Default `false`
```
If possible should be some comments explaining when these happen such as recoil, status, weather, etc

Tempo-anon commented on src/data/battler-tags.ts:
```diff
@@ -1872,8 +1871,12 @@ export class PerishSongTag extends BattlerTag {
+        result: HitResult.ONE_HIT_KO,
+        ignoreSegments: true,
+        preventEndure: true,
+        ignoreDynamaxReduction: true,
```
A much more elegant way of handling it. Are there any other times when this should be true though?

DayKev replied:
All OHKO instances most likely. I'll do a quick check.

DayKev replied:
Added it to all sources of ratio-based damage (ie: Steel Beam should deal half the user's HP in damage, this shouldn't be reduced by the dynamax damage reduction) and OHKOs.

I wonder if it would be better to just implement the dynamax HP boost as an actual HP boost instead of a damage reduction. You could add a multiplier into the code that calculates a pokemon's HP if that pokemon is dynamaxed, and then whenever a pokemon dynamaxes or reverts from dynamax update their hp and call `pokemon.calculateStats()` to update the UI.

Tempo-anon replied:
I thought modifying HP would have been more complex than just a damage modifier but maybe I was incorrect :joltdead:

DayKev committed:
Update tsdocs of `Pokemon.damage`

DayKev committed:
Change `if` check for sturdy/endure to be clearer

DayKev committed:
Add `ignoreDynamaxReduction: true` to all ratio-based damage and OHKOs

DayKev committed:
Consolidate `pokemon.turnData.damageTaken += ...` to `Pokemon.damage()`

torranx commented on src/field/pokemon.ts:
```diff
@@ -3045,91 +3046,134 @@ export abstract class Pokemon extends Phaser.GameObjects.Container {
+      ignoreFaintPhase = false,
+      source,
+      ignoreDynamaxReduction = false,
+    }: {
+      result?: DamageResult;
+      isCritical?: boolean;
+      ignoreSegments?: boolean;
+      preventEndure?: boolean;
+      ignoreFaintPhase?: boolean;
+      source?: Pokemon;
+      ignoreDynamaxReduction?: boolean;
+    } = {},
```
nit: move this into an interface (and the other params)

DayKev replied:
A single-use interface that has to be defined away from the function isn't necessary.

torranx replied:
I agree but it improves readability

DayKev replied:
I forgot about the in-editor tsdocs viewing, and this does improve that, so I added it.

DayKev committed:
Create `src/@types/DamageFunctionOptions.ts`

MokaStitcher commented on src/field/pokemon.ts:
```diff
@@ -3046,91 +3048,126 @@ export abstract class Pokemon extends Phaser.GameObjects.Container {
+      ignoreSegments = false,
+      preventEndure = false,
+      ignoreFaintPhase = false,
+      source,
```
feels weird to have a param with no default value between 2 params with a default value

MokaStitcher commented on src/data/ab-attrs/form-block-damage-ab-attr.ts:
```diff
@@ -54,7 +54,12 @@ export class FormBlockDamageAbAttr extends ReceivedMoveDamageMultiplierAbAttr {
         multiplier.value *= this.multiplier;
```
I can't comment on the specific line, but the tsdoc for this function seems outdated, do you mind updating it? (`args` should be replaced by `multiplier`)

MokaStitcher replied:
I guess that's a valid way to not make it outdated :plains:

MokaStitcher reviewed (commented):
Can you update the description to explain the introduction of the `ignoreDynamaxReduction` parameter, just so that we have a trace that doesn't require reading through the discussion

otherwise, not playtested yet but lgtm

MokaStitcher commented:
(new conflict to solve)

Also I seem to be getting decimal damage numbers in some cases (here, burn damage)
![image](https://github.com/user-attachments/assets/beedcf43-f633-4865-a71c-98c8b154dc2c)
![image](https://github.com/user-attachments/assets/33644900-b2f6-4ad1-bf19-d0779e6de4cb)

DayKev committed:
Re-order `damageAndUpdate` params

DayKev committed:
Update tsdocs for `FormBlockDamageAbAttr`

DayKev committed:
Merge branch 'beta' into refactor/damage-functions

DayKev committed:
Fix damage values in `PostTurnStatusEffectPhase`

DayKev commented:
> Also I seem to be getting decimal damage numbers in some cases (here, burn damage)

Found the issue, when I replaced the bitshifts with normal division operators in `PostTurnStatusEffectPhase` that revealed that the damage values weren't wrapped in `toDmgValue()` so I added them in.

DayKev commented:
> > Also I seem to be getting decimal damage numbers in some cases (here, burn damage)
> 
> Found the issue, when I replaced the bitshifts with normal division operators in `PostTurnStatusEffectPhase` that revealed that the damage values weren't wrapped in `toDmgValue()` so I added them in.

Actually I think maybe that shouldn't even be necessary, it should probably be handled by `Pokemon.damageAndUpdate()`. Maybe I'll do that as a follow-up to this PR.

DayKev committed:
Add `TODO` in `post-turn-status-effect-phase.ts` on animation code

DayKev committed:
Merge branch 'beta' into refactor/damage-functions

DayKev committed:
Merge branch 'beta' into refactor/damage-functions

DayKev committed:
Merge branch 'beta' into refactor/damage-functions

DayKev committed:
Merge branch 'beta' into refactor/damage-functions

DayKev committed:
Fix merge issues

DayKev committed:
Fix GMax OHKO damage

DayKev committed:
Merge branch 'beta' into refactor/damage-functions

flx-sta committed:
Merge branch 'beta' into refactor/damage-functions

Tempo-anon committed:
Merge branch 'beta' into refactor/damage-functions

PigeonBar commented on src/data/battler-tags.ts:
```diff
@@ -981,7 +984,7 @@ export class SeedTag extends BattlerTag {
           );
 
-          const damage = pokemon.damageAndUpdate(toDmgValue(pokemon.getMaxHp() / 8));
+          const damage = pokemon.damageAndUpdate(toDmgValue(pokemon.getMaxHp() / 8), { ignoreDynamaxReduction: true });
```
![20250322 Gmax Percent Damage](https://github.com/user-attachments/assets/e311f31b-3d59-4c44-b498-e1640bd8bb28)

According to Showdown and Bulbapedia, effects that deal a percentage of a G-max Pokemon's health still follow dynamax damage reduction. (G-max Grimmsnarl took around 1/16 instead of 1/8 of its HP in damage from each of Leech Seed and Wrap.)

DayKev replied:
GameFreak please...

DayKev replied:
Does this include things like Substitute? Is there even a way for a Dynamax'd pokemon to use Substitute?

PigeonBar replied:
I think Dynamax'd Pokemon in mainline are unable to use any moves other than Max Moves, so stuff like Substitute or Steel Beam would not be testable in mainline.

DayKev replied:
What should we do about that?

PigeonBar commented on src/field/pokemon.ts:
```diff
@@ -3112,91 +3114,126 @@ export abstract class Pokemon extends Phaser.GameObjects.Container {
     const healAmount = Math.min(amount, this.getMaxHp() - this.hp);
     this.hp += healAmount;
+    if (!quiet && this.isOnField()) {
+      globalScene.damageNumberHandler.add(this, amount, HitResult.HEAL);
```
It turns out `healAmount` should be used here instead of `amount`, so that the displayed heal number does not show the Pokemon healing to more than full HP.

DayKev replied:
Put up a draft PR #867 with fixes to the issues noticed so far.
